[tasks] Add auto runtime initialization template to LibraryBuilder

mdh1418 opened the pull request:
Contributes to https://github.com/dotnet/runtime/issues/79377

Incorporates mono's lazy runtime init capability https://github.com/dotnet/runtime/pull/82253 into the library builder task https://github.com/dotnet/runtime/pull/81919.

In the target use case of having a self-contained library containing the mono runtime with custom managed code, the mono runtime needs to be initialized when the library is loaded into the native application. In order to do so, one can either

A. leverage the default `runtime-init-callback` and upon loading the shared library, rely on the auto initialization logic in `autoinit.c` (Introduced by this PR) to setup the callback via `mono_set_runtime_init_callback`. The callback is set when first loading the shared library due to the constructor.

or 

B. pass in a custom `runtime-init-callback=` introduced in https://github.com/dotnet/runtime/pull/82253 and have one's own runtime initialization startup logic instead of `autoinit.c`.

--------------

The runtime initialization logic looked at https://github.com/dotnet/runtime/blob/main/src/tasks/AndroidAppBuilder/Templates/monodroid.c for inspiration.
In attempt to have a barebones version, the following is done:

- Set the path to assemblies, expected in environment variable `DOTNET_ASSETS_PATH`
- Initialize possible runtime config file
- Initialize app context environment variables
- register all aot modules
- <s>register bundled modules</s> // TODO when bundling is implemented
- set the path to search for assemblies based on environment variable `DOTNET_ASSETS_PATH` or custom defined variable
- set to aot only // For now until Mixed mode is worked out
- <s>add a custom assembly preload hook</s> // No longer needed as assembly paths are set
- <s>allow for debugging</s> //TODO in followup
- <s>add install hook for aot data</s> // Not default scenario
- add signal chaining
- initialize the mono runtime
- load all assemblies that contain exported symbols

---------------

Example generated files
[CMakeLists](https://gist.github.com/mdh1418/c50de25dfe0e4809a1bf21693d46483d)
[shared_library_log.h](https://gist.github.com/mdh1418/6ef3cc1ce9b3f50ab5a94d52c79b67c5)
[autoinit.c](https://gist.github.com/mdh1418/195e11a0b62b5f4a5ba37579056bd254)
[load_assemblies.c](https://gist.github.com/mdh1418/9917ded440410b9cac5598e5f7fae088)

mdh1418 committed:
[tasks] Add auto runtime initialization template to LibraryBuilder

mdh1418 committed:
Grab shared library directory

mdh1418 committed:
[tasks] Extend auto initialization template to account for customization

mdh1418 committed:
Revert "Grab shared library directory"

This reverts commit bc86f2b91c76f0e40ad60516dcb1031d02e7054b.

steveisok commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -226,6 +253,79 @@ private static void WriteLinkerScriptFile(string exportsFile, List<string> expor
+            return "    return;";
+
+        var runtimeConfig = new StringBuilder();
+        runtimeConfig.Append($"    char *file_name = {RuntimeConfigBinFile};");
```
I'm not sure why this is done here?

steveisok replied:
We should confirm if we always get a runtimeconfig file. I can't remember if we do or not.

mdh1418 replied:
Right, I added this for now because I'm not familiar with that file and what it does, so added this for flexibility for now.

steveisok replied:
I think we can assume that runtimeconfig.bin will be next to the assemblies, so I think this can be moved back into autoinit.c.  

I think we are going to have to run the `_AndroidGenerateRuntimeConfig` target before the library builder runs to make sure we have a bin file. The appbuilder targets do this currently because it's assumed they are building every piece of the app.

mdh1418 replied:
I moved the `monovm_runtimeconfig_initialize` logic and `monovm_initialize` logic back into `autoinit.c`, with the only substitution being the appctx value for `RUNTIME_IDENTIFIER`, which has `RuntimeIdentifier` passed to it. 

Both of these will rely on the path to where assemblies are located (currently `bundle_path`) being set, which is done at the beginning of the runtime initialization callback.

The runtimeconfig file that is parsed in `monovm_runtimeconfig_initialize` has consistenly been named `runtimeconfig.bin`. Unless we anticipate the naming to vary (Android/Apple/Wasm/Wasi have it as `runtimeconfig.bin`) I dont think we *need* the `_AndroidGenerateRuntimeConfig` to run before `_BuildNativeLibrary` because the only dependency is that the actual file is in the "bundle_path" when the runtime is intialized.

steveisok commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -226,6 +253,79 @@ private static void WriteLinkerScriptFile(string exportsFile, List<string> expor
+        int numArgs = AppContextKeys.Length;
+        var appContextEnvVariables = new StringBuilder();
+
+        appContextEnvVariables.AppendLine($"    const char **appctx_keys, **appctx_values = (char**)malloc({numArgs} * sizeof(char*));");
```
Same, why can't we do this in autoinit.c?

mdh1418 replied:
I added this in LibraryBuilder because the amount of memory to allocate depends on the number of variables.

Or would it be preferred to have something like 
autoinit.c
```
    const char **appctx_keys, **appctx_values = (char**)malloc({%NUM_ARGS%} * sizeof(char*));
    ...
    monovm_initialize(%NUM_ARGS%, appctx_keys, appctx_values);
```

LibraryBuilder.cs
```
    private void WriteAutoInitializationFromTemplate()
    {
        string appContextEnvVariables = GenerateAppContextEnvVariables();
        string assembliesLoader = GenerateAssembliesLoader();
        string assetsPath = GenerateAssetsPath();
        string runtimeConfig = GenerateRuntimeConfig();
        File.WriteAllText(Path.Combine(OutputDirectory, "autoinit.c"),
            Utils.GetEmbeddedResource("autoinit.c")
                .Replace("%NUM_ARGS%", AppContextKeys.Length) // And checking AppContextKeys.Length != AppContextValues.Length much earlier
        ...
```

steveisok replied:
@lateralusX does nativeaot allow extra app context values like this?

lateralusX replied:
Looking through the bootstrapping code, it doesn't look like its using the app context properties there, the bootstrapper pretty much calls into RhInitialize that calls InitDLL and that setup the runtime. Having that said it looks like they run __managed__Startup as part of library mode that does some additional initialization in managed code added by IL compiler:

compilationRoots.Add(new MainMethodRootProvider(entrypointModule, CreateInitializerList(typeSystemContext)));
compilationRoots.Add(new RuntimeConfigurationRootProvider(_runtimeOptions));
compilationRoots.Add(new ExpectedIsaFeaturesRootProvider(instructionSetSupport));

Looks like the initializer list also add app context settings into a init method that gets run at this stage and app context switches is arguments passed to IL compiler:

// If there are any AppContext switches the user wishes to enable, generate code that sets them.
if (_appContextSwitches.Count > 0)
{
    MethodDesc appContextInitMethod = new Internal.IL.Stubs.StartupCode.AppContextInitializerMethod(
    context.GeneratedAssembly.GetGlobalModuleType(), _appContextSwitches);
    initializerList.Add(appContextInitMethod);
}

and argument examples to IL compiler:

<IlcArg Include="@(RuntimeHostConfigurationOption->'--appcontextswitch:%(Identity)=%(Value)')" />
<IlcArg Include="--appcontextswitch:RUNTIME_IDENTIFIER=$(RuntimeIdentifier)" />

So you can set app context properties through IL compiler arguments. I guess we can do the same as msbuild arguments passed down to library builders from our targets.

mdh1418 replied:
Changed to not allow any extra app context entries for now. We can add more if there is a need to install more environment variables that aren't passable into the generated runtimeconfig.json.

steveisok commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,162 @@
+
+    mono_set_signal_chaining (true);
+
+    mono_jit_init ("dotnet.android"); // Pass in via LibraryBuilder?
```
Where possible, we need this to run on both ios and android.

steveisok replied:
I don't know how important the label truly is.  Since this could be side by side, we need something somewhat unique.

mdh1418 replied:
I dont think it will matter too much, and `mono.self.contained.library` is probably unique enough.

steveisok commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -226,6 +253,79 @@ private static void WriteLinkerScriptFile(string exportsFile, List<string> expor
+        var assembliesLoader = new StringBuilder();
+        foreach (string exportedAssembly in exportedAssemblies)
+        {
+            assembliesLoader.AppendLine($"    mono_assembly_open(\"{exportedAssembly}\", NULL);");
```
Since loading the assembly first makes sure everything initializes properly, having a function would give the same ability to custom initializers.  On the other hand, this makes things much more smooth.  Come to think of it, why don't we have a function that just does what you're doing in assembly_list.c?

mdh1418 replied:
We could move the `mono_assembly_open` logic in `assembly_list.c` (and probably rename it to like `load_assemblies`), that would just be refactoring the code if it helps to have that separation. On the other hand, if there are no assemblies to load (no UnmanagedCallersOnly decorated functions) then that function would be an empty function yet still be called at the end of `runtime_init_callback`.

What do we anticipate as the number of assemblies that the end user will need to be loaded at the end of the runtime initialization? Is it in the order of 1-10 or 10-50? If its on the larger side, I think having the `mono_assembly_open` in a separate file could be a clear separation, but does the separation need to be in 2 files?

mdh1418 replied:
Moved `mono_assembly_open` into the separate function (now called `load_assemblies_with_exported_symbols`) in the separate source file (now called `load_assemblies.c`) and replaced the previous logic that generated a function that returned a list of assemblies.

The change is specifically in the commit 1Move assembly loading into separate function`
https://github.com/mdh1418/runtime/commit/3b71e955501c32489a01dfe503e40898da367152

mdh1418 committed:
Remove platform specific naming

steveisok commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -66,6 +66,26 @@ public bool IsSharedLibrary
+    /// </summary>
+    public bool UsesCustomRuntimeInitCallback { get; set; }
+
+    public string? AssetsPath { get; set; }
```
`AssetsPath` seems too generic. Maybe we should rename this to AssemblyLocationPath?  I'm open to suggestions - I can get lost in naming with the best of them.

mdh1418 replied:
Changed to `AssembliesLocation`

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,162 @@
+void
+init_mono_runtime ()
+{
+    bundle_path = monoeg_g_getenv("%ASSETS_PATH%");
```
Maybe this should be done later, the library can be loaded very early and embedder doesn't have had the chance to set anything yet in the env variable.

lateralusX replied:
If this is the only call into eglib, then maybe you should just use the underlying getenv instead, eglib symbols go through a remap process, that is why they are named monoeg_ so there is a file in runtime eglib.remap.h that pretty much do a #define g_getenv monoeg_g_getenv, and that is optional, so in theory the symbol might not be remapped, but not sure if that is ever used anymore, but the option is still available when building runtime.

mdh1418 replied:
Good point, moved the environment variable getter into the callback itself as the first thing it does, and leveraged `getenv` directly instead of needing to import `eglib` for just one API.

Yeah.... I realized that was what the symbol was exported as when I couldn't get `g_getenv` working even when adding in a prototype to the beginning of autoinit.c

Should I be adding prototypes for all mono api's? 
`monovm_runtimeconfig_initialize`
`monovm_initialize`
`mono_assembly_open`
`mono_assembly_name_get_name`
`mono_assembly_name_get_culture`
`mono_assembly_get_name`
`mono_assembly_name_get_name`
`mono_set_assemblies_path`
`mono_jit_set_aot_only`
`mono_install_assembly_preload_hook`
`mono_install_load_aot_data_hook`
`mono_set_signal_chaining`
`mono_jit_init`
`mono_set_runtime_init_callback`

mdh1418 replied:
Should I make a separate `autoinit.h` header as well to hold all of these prototypes?

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,162 @@
+}
+
+void
+init_mono_runtime ()
```
I think this method should have a different name, you get the impression that its initialize the runtime, but it just sets up the callback hook, you should just do that in the autoinit constructor function directly.

mdh1418 replied:
Removed this intermediate wrapper

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,162 @@
+
+static char *bundle_path;
+
+void register_aot_modules (void);
```
Shouldn't we generate the registration code of aot modules direct in this file when building library instead of creating multiple C source files?

steveisok replied:
The aot compiler generates the c file on our behalf. We could read that file in and place the contents here.

mdh1418 replied:
Kept the aot module registration file (modules.c) generation in the MonoAOTCompiler as it gets placed in the Output Directory. The CMakeLists.txt of the LibraryBuilder is also generated in the same Output Directory, adds modules.c as an extra source (DOTNET_EXTRA_SOURCES), and links it  into the static lib `libaot_library.a`

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,162 @@
+
+    mono_install_assembly_preload_hook (mono_assembly_preload_hook, NULL);
+
+    // TODO test debug scenario
```
I think we should remove this until we implement debugger support in library builder.

mdh1418 replied:
Removed

mdh1418 committed:
[tasks] Move App Context setup back to autoinit.c

mdh1418 committed:
Remove registering bundled modules and debugging until they can be validated

mdh1418 committed:
Remove unnecessary wrapper

mdh1418 committed:
Grab bundle_path later without eglib

Setting bundle_path in constructor may be too early for environment variables to be set
so grab it first thing when initializing the runtime.

Use getenv directly instead of calling into eglib with just one api

mdh1418 committed:
Rename AssetsPath to AssembliesLocation

mdh1418 committed:
Move assembly loading into separate function

mdh1418 committed:
Remove unnecessary function

mdh1418 committed:
Cleanup CMakeLists template indentation

mdh1418 committed:
Add dotnet header to autoinit.c

mdh1418 committed:
Default bundle_path if assemblies path environment variable not set

mdh1418 committed:
Pass bundle_path to load assemblies

mdh1418 committed:
Add logging and abort on failures

lambdageek commented on src/tasks/LibraryBuilder/Templates/load_assemblies.c:
```diff
@@ -1,9 +1,27 @@
+
+    assert (num_char > 0 && num_char < str_len);
+
+    mono_assembly_open (assembly_path, NULL);
```
oh, @mdh1418 I realized I was wrong yesterday - `mono_assembly_open` works on assembly file paths; `mono_assembly_load` (which takes a `MonoAssemblyName*`) works with assembly names.

So the fact that you use `mono_set_assemblies_path` won't help `mono_assembly_open`, but it will help `mono_assembly_load` (and the runtime internals that also do assembly probing by paths).

mdh1418 replied:
Ah, will `mono_assembly_load` also load the GOT slots? I'm not familiar with the differences yet, will have to look.
It also looks like `mono_assembly_load` takes in a basedir, but not sure how it gets used quite yet, maybe `mono_assembly_load_with_partial_name` might do the trick, will have to try it out.

mdh1418 replied:
`mono_assembly_load_with_partial_name` worked!

steveisok commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -256,23 +256,42 @@ private static void WriteLinkerScriptFile(string exportsFile, List<string> expor
                 .Replace("%RUNTIME_IDENTIFIER%", RuntimeIdentifier));
     }
 
+    private string GenerateLog(string level)
```
My feeling is that it's better to `#ifdef` these in autoinit.

mdh1418 replied:
Used `#ifdef` instead and passed `-DHOST_ANDROID` into the CMakeLists `if(TARGETS_ANDROID)` block

mdh1418 committed:
Load assemblies with partial name

steveisok commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -66,6 +66,26 @@ public bool IsSharedLibrary
+
+    /// <summary>
+    /// </summary>
+    public string? RuntimeConfigBinFile { get; set; }
```
I don't think this is needed any longer.

mdh1418 replied:
Thanks for the catch!

mdh1418 committed:
Clean up targets and unneeded properties

mdh1418 committed:
Rename source file

mdh1418 commented:
Can I get another set of reviews @steveisok @lateralusX @lambdageek @akoeplinger ?

steveisok commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -113,6 +126,14 @@ public override bool Execute()
+            extraSources.AppendLine("    autoinit.c");
+        }
+
+        GenerateAssembliesLoader();
```
if exportedAssemblies == 0, we should error out the task.  Building a library without exports wouldn't be usable.

mdh1418 replied:
good catch, added a throw inside `GatherAotSourcesObjects` right after looping over the assemblies to throw if no exported symbols were found.

steveisok reviewed (approved):
Outside of my last feedback item, looks good!

mdh1418 committed:
Remove preload hook and make out of memory error more generic

mdh1418 committed:
Throw when no exported symbols are found

mdh1418 committed:
Remove aot data load prehook and aotdata generation

lateralusX commented on src/mono/msbuild/android/build/AndroidApp.targets:
```diff
@@ -151,6 +151,7 @@
 
     <PropertyGroup>
```
@steveisok Does library builder have a dependency on these targets, meaning they will need to go into workload? If so, maybe we should, make the naming better to suite the different scenarios and drop the "App" part in file name + used targets, since it is now not only used to build apps anymore.

steveisok replied:
Yes, I would prefer this be renamed in the workload PR and not in this one,

lateralusX replied:
Agree, just wanted to highlight it, do we have a workload PR up yet?

steveisok replied:
I have it in a branch. I'll throw post a PR today.

lateralusX commented on src/mono/msbuild/apple/build/AppleApp.targets:
```diff
@@ -93,6 +93,7 @@
 
     <PropertyGroup>
       <_EnableUnmanagedCallersOnlyMethodsExport Condition="'$(_IsLibraryMode)' == 'true'">true</_EnableUnmanagedCallersOnlyMethodsExport>
+      <_UseAotDataFile Condition="'$(_IsLibraryMode)' == 'true' and '_UseAotDataFile' == ''">false</_UseAotDataFile>
```
Shouldn't the apple targets have same features as the Android? Missing the callback compiler arguments needed for library mode.

mdh1418 replied:
Good catch. I didn't focus on fleshing out the apple side since we haven't yet finished getting the LibraryBuilder working for the Apple side anyways.
Would it be better to just add the Apple side of things when we get things rolling there (and do the same for wasm/wasi when we start on that front + other platforms)?
I'll go ahead and add all the pieces for Apple

lateralusX replied:
Yes, add at least the platform we know users will start out testing.

lateralusX commented on src/tasks/LibraryBuilder/Templates/CMakeLists.txt.template:
```diff
@@ -5,21 +5,20 @@ project(%LIBRARY_NAME%)
 set(DOTNET_EXTRA_SOURCES
-    %ExtraSources%
-    assembly_list.c
+%ExtraSources%    preloaded_assemblies.c
```
put them on separate lines to increase readability.

mdh1418 replied:
Initially I placed it as such so the resulting CMakeFile after replacing `%ExtraSources%` wouldn't have weird spaces or extra empty lines.

I'll rearrange it as

```
set(DOTNET_EXTRA_SOURCES
    preloaded_assemblies.c
%ExtraSources%)
```

so if `%ExtraSources%` is empty
```
set(DOTNET_EXTRA_SOURCES
    preloaded_assemblies.c
)
```
and otherwise
```

set(DOTNET_EXTRA_SOURCES
    preloaded_assemblies.c
    extrasource_1
    extrasource_2
    ...
    extrasource_n
)
```

lateralusX commented on src/tasks/LibraryBuilder/Templates/CMakeLists.txt.template:
```diff
@@ -35,8 +34,9 @@ add_library(
+        log
     )
+    add_definitions(-DHOST_ANDROID=1)
 else()
```
Should we be more precise on the platforms we support? And have a else case triggering an error with unsupported platforms?

mdh1418 replied:
We have a guard in the LibraryBuilder task that checks for the platforms TargetOS. Right now its `"android" or "ios" or "tvos" or "maccatalyst" => true`. I can throw a `send_message(FATAL_ERROR...)` in the else guard as well.

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+        arg->runtimeconfig.name.path = file_path;
+        monovm_runtimeconfig_initialize (arg, cleanup_runtime_config, file_path);
+    } else {
+        LOG_INFO ("Could not stat file '%s'. Runtime configuration properties not initialized.\n", file_path);
```
Maybe this is a little to technical for users, "stat file", maybe its better to say that "Could not find file"

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+static void
+cleanup_runtime_config (MonovmRuntimeConfigArguments *args, void *user_data)
+{
+    free (args);
```
free arg->runtimeconfig.name.path as well?

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+        MonovmRuntimeConfigArguments *arg = (MonovmRuntimeConfigArguments *)malloc (sizeof (MonovmRuntimeConfigArguments));
+        arg->kind = 0;
+        arg->runtimeconfig.name.path = file_path;
+        monovm_runtimeconfig_initialize (arg, cleanup_runtime_config, file_path);
```
Fix cleanup_runtime_config to fee file_path from arg instead, and pass NULL as user_data here.

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+    struct stat buffer;
+
+    if (stat (file_path, &buffer) == 0) {
+        MonovmRuntimeConfigArguments *arg = (MonovmRuntimeConfigArguments *)malloc (sizeof (MonovmRuntimeConfigArguments));
```
check for alloc failure.

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+        abort ();
+    }
+
+    int num_char = snprintf (file_path, str_len, "%s/%s", bundle_path, file_name);
```
check for snprintf success or error.

mdh1418 replied:
Because `file_path` exists at this point (would have aborted above if couldn't allocate the needed memory) and `str_len` is constructed to be longer than the length of the formatted data string, then wouldn't the only point of failure be an encoding error? How would the users normally go about resolving encoding error?

Adding 
```
if (num_char < 0)
    LOG_ERROR ("Encoding error while formatting '%s' and '%s' into \"%%s/%%s\".\n", bundle_path, file_name);
```

lateralusX replied:
Just check if it succeeds or fails, consider success if num_char > 0 && num_char < str_len.

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+    size_t str_len = sizeof (char) * (strlen (bundle_path) + strlen (file_name) + 2); // +1 "/", +1 null-terminating char
+    char *file_path = (char *)malloc (str_len);
+    if (!file_path) {
+        LOG_ERROR ("Out of memory.\n");
```
maybe you should put abort in LOG_ERROR macro, since it will always abort on error.

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+void
+runtime_init_callback ()
+{
+    const char *assemblies_path = strdup(getenv("%ASSEMBLIES_LOCATION%"));
```
getenv returns NULL if variable is not found.

mdh1418 replied:
Good catch, 
changed to 
```
    const char *assemblies_path = getenv("%ASSEMBLIES_LOCATION%");
    if (assemblies_path && assemblies_path[0] != '\0')
        bundle_path = strdup (assemblies_path);
    else
        bundle_path = strdup ("./");
    if (!bundle_path)
        LOG_ERROR ("Out of memory.\n");
```

and freeing `bundle_path` at the end of the init callback instead of `assemblies_path` since [It isn't safe to modify the value of the environment variable using the returned pointer.](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/getenv-wgetenv?view=msvc-170)

mdh1418 replied:
After seeing https://github.com/dotnet/runtime/pull/83050#discussion_r1136792693
```
    bundle_path = getenv("%ASSEMBLIES_LOCATION%");
    if (assemblies_path && assemblies_path[0] != '\0')
        bundle_path = "./";
```
and no need to free afterwards.

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+        LOG_ERROR ("Out of memory.\n");
+        abort ();
+    }
+    bundle_path = (assemblies_path[0] != '\0') ? assemblies_path : "./";
```
Maybe you should put all scenarios where we can fail to have an assemblies_path and fallback to default path instead?

lateralusX replied:
Maybe add a LOG_INFO with the bundle_path we end up using.

mdh1418 replied:
Oh right, `bundle_path` doesn't need to by a copy of `assemblies_path` obtained from the env variable since we don't modify it. I'll just set `bundle_path` to `getenv (%ASSEMBLIES_LOCATION%)` and if thats `null`, I'll default it to `"./"`

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+#ifdef HOST_ANDROID
+#include <android/log.h>
+
+#define LOG_INFO(fmt, ...) __android_log_print(ANDROID_LOG_INFO, "MONO_SELF_CONTAINED_LIBRARY", fmt, ##__VA_ARGS__)
```
Do you use LOG_INFO? I think library init must be quite if we don't hit errors.

mdh1418 replied:
Right now, its only being used to notify users that runtime config file could not be found in case they dont see app context properties they were expecting to exist from the built runtime config(monodroid.c and runtime.m both continue even if its not found). It didnt feel like a scenario in which we should error, but should we?

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+
+#define LOG_INFO(fmt, ...) __android_log_print(ANDROID_LOG_INFO, "MONO_SELF_CONTAINED_LIBRARY", fmt, ##__VA_ARGS__)
+#define LOG_ERROR(fmt, ...) __android_log_print(ANDROID_LOG_ERROR, "MONO_SELF_CONTAINED_LIBRARY", fmt, ##__VA_ARGS__)
+#else
```
I think we should be specific with IOS targets here and not rely on else fallback. Maybe we even should do a #error in the else case to give compile error for unsupported HOSTS.

mdh1418 replied:
Added the definition for apple mobile platforms and used an else to throw an error for unsupported hosts

lateralusX commented on src/tasks/LibraryBuilder/Templates/preloaded_assemblies.c:
```diff
@@ -0,0 +1,30 @@
+
+#include <mono/metadata/assembly.h>
+
+#ifdef HOST_ANDROID
```
Maybe you should put this into a shared template header so we only have it in one place? If you decide to keep it here then look at comments for the same code above.

mdh1418 replied:
Added to a shared template header that is named `shared_library_log.h` (until we need to add more stuff and the name needs to be changed)

lateralusX commented on src/tasks/LibraryBuilder/Templates/preloaded_assemblies.c:
```diff
@@ -0,0 +1,30 @@
+void
+load_assemblies_with_exported_symbols ()
+{
+%ASSEMBLIES_LOADER%}
```
put } on a new line.

lateralusX commented on src/tasks/LibraryBuilder/Templates/preloaded_assemblies.c:
```diff
@@ -0,0 +1,30 @@
+#endif
+
+void
+mono_assembly_load_with_partial_name_check (const char* filename)
```
this is only internal for this source file, correct? If so make it static and drop mono prefix.

lateralusX commented on src/tasks/LibraryBuilder/Templates/preloaded_assemblies.c:
```diff
@@ -0,0 +1,30 @@
+}
+
+void
+load_assemblies_with_exported_symbols ()
```
preload_assemblies_with_exported_symbols ?

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+    free ((void *)assemblies_path);
+}
+
+void __attribute__((constructor))
```
Does this need to be "public" symbol in order for the constructor to work? If that is the case, can we still hide it from being public out of the build shared library? If not, we would need to name it uniquely or it will collide with other shared libraries loaded into the same process.

steveisok replied:
No, all symbols except the ones in the exported symbols files are private via the linker script.

mdh1418 commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,116 @@
+        abort ();
+    }
+
+    load_assemblies_with_exported_symbols ();
```
yes, forgot to change this function after changing the file name.

mdh1418 committed:
Address feedback

lateralusX commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -113,6 +126,17 @@ public override bool Execute()
         GatherAotSourcesObjects(aotSources, aotObjects, extraSources, linkerArgs);
         GatherLinkerArgs(linkerArgs);
 
+        File.WriteAllText(Path.Combine(OutputDirectory, "shared_library_log.h"),
```
I think we should just name this library_builder.h and be done.

lateralusX replied:
or use - as delimitator since we use that for other header files, so library-builder.h

lateralusX commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -145,10 +168,14 @@ private void GatherAotSourcesObjects(StringBuilder aotSources, StringBuilder aot
         }
+        if (exportedAssemblies.Count == 0)
+        {
+            throw new LogAsErrorException($"None of the compiled assemblies contain exported symbols. Resulting shared library would be unusable.");
```
Is this only for the shared library scenario or the static one as well? Should we really limit this? You could build a library and handle all embedding on your own and still have a lot of value with library builder producing a self-contained library for you.

mdh1418 replied:
Didn't realize that the static library version would be okay without any exported methods from `UnmanagedCallersOnly` decorated managed methods. I think we figured that if there were no native-to-managed wrappers from calling exported functions then no `mono_threads_attach_coop` would be called to try to initialize the runtime. For those building and working with static library, would they be able to initialize the runtime without going through that route? 

A general question, would it be obvious how to handle the embedding and have a call to initialize the runtime? Should we have some doc to walkthrough what would need to be done for both scenarios of shared or static built?

I'll add in `&& IsSharedLibrary` to the condition to make it shared library specific.

lateralusX replied:
So you have different scenarios:

1. Build a shared library, since that will only export unmanagedcallers entrypoint the library can't be used unless we have those exports and in that case we need to use the runtime_init_callback default or custom.
2. Build a static library, now all symbols exported from runtime are still public as well as symbols exported using unmanagedcallers entrypoint. User might still want to have the runtime_init_callback as default or custom since they don't want to initialize the runtime, just call the entrypoints exported using unmangedcallers entrypoint.
3. Build a static library, but since we have all runtime API's available, do full runtime init on your own, before calling any entrypoints or using mono_runtime_invoke to call any managed method, so no need to unmanagedcallers entry points and user don't want any auto init logic in this scenario, so needs to be turned off. Quite possible that this mode will be used by iOS/Android SDK to get total control of runtime init as well as how the library will be packaged and distributed.

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,104 @@
+
+static const char *bundle_path;
+
+void register_aot_modules (void);
```
Since you now have a header, I think you should move the prototypes into that header instead.

mdh1418 replied:
I originally was thinking that if users had their own runtime callback, then the prototypes would be unneeded (yet they would still be linked in via the header file). But on second thought, they would still need to invoke these right? Or are there ways to initialize the runtime without `register_aot_modules` with AOT on?

lateralusX commented on src/tasks/LibraryBuilder/Templates/autoinit.c:
```diff
@@ -0,0 +1,104 @@
+        arg->runtimeconfig.name.path = file_path;
+        monovm_runtimeconfig_initialize (arg, cleanup_runtime_config, NULL);
+    } else {
+        LOG_INFO ("Could not find file '%s'. Runtime configuration properties not initialized.\n", file_path);
```
I think library init needs to be quite , if not having a runtimeconfig is a valid option that can happen, then we shouldn't log anything and if its not we should error out, but I think we should just accept that case and don't log anything.

mdh1418 replied:
Removed LOG_INFO and removed from the header file.

lateralusX commented on src/mono/msbuild/android/build/AndroidApp.targets:
```diff
@@ -85,6 +85,8 @@
       <MonoAOTCompilerDefaultAotArguments Include="static" />
       <MonoAOTCompilerDefaultAotArguments Include="dwarfdebug" />
       <MonoAOTCompilerDefaultAotArguments Condition="'$(_IsLibraryMode)' == 'true'" Include="direct-icalls" />
+      <MonoAOTCompilerDefaultAotArguments Condition="'$(_IsLibraryMode)' == 'true' and $(CustomRuntimeInitCallback) == ''" Include="runtime-init-callback" />
```
I think it would make sense to have the option to opt out of runtime-init-callback and handle of of the runtime init when building a static library. Since we will hide all symbols for shared library we need to do the runtime-init-callback, so now way to opt out of it in that mode, but for static library we should have the option.

mdh1418 replied:
Added an option `UsesRuntimeInitCallback` that users can set to not `true` in order to omit any `runtime-init-callback*` args being passed to the mono compiler.

So `UsesRuntimeInitCallback` and `CustomRuntimeInitCallback` can be used in conjunction to let the compiler know whether or not a `runtime-init-callback` is needed, and if it is, whether or not it should be the default or the custom passed in via `CustomRuntimeInitCallback`. Not sure what the best way right now to surface those properties to the public. I suppose once we have a doc to demonstrate how all of the piece fit together, we can surface it there?
```
    <PropertyGroup Condition="'($_IsLibraryMode)' == 'true'">
      <_UsesRuntimeInitCallback>$(UsesRuntimeInitCallback)</_UsesRuntimeInitCallback>
      <_UsesRuntimeInitCallback Condition="'$(_UsesRuntimeInitCallback)' == ''">true</_UsesRuntimeInitCallback>
    </PropertyGroup>
    
    <ItemGroup>
      ...
      <MonoAOTCompilerDefaultAotArguments Condition="'$(_UsesRuntimeInitCallback)' == 'true' and $(CustomRuntimeInitCallback) == ''" Include="runtime-init-callback" />
      <MonoAOTCompilerDefaultAotArguments Condition="'$(_UsesRuntimeInitCallback)' == 'true' and $(CustomRuntimeInitCallback) != ''" Include="runtime-init-callback=$(CustomRuntimeInitCallback)" />
    </ItemGroup>
```

lateralusX replied:
Yes we surface them when needed, by default it should be using runtime init callback and it should use our default implementation. Turning off UsesRuntimeInitCallback should also omit creating autoinit.c since that won't be needed.

mdh1418 commented on src/tasks/LibraryBuilder/Templates/preloaded_assemblies.c:
```diff
@@ -0,0 +1,21 @@
+{
+    MonoAssembly *assembly = mono_assembly_load_with_partial_name (filename, NULL);
+    if (!assembly)
+        LOG_ERROR ("Could not open assembly '%s'. Unable to properly initialize GOT slots.\n", filename);
```
Good point 😅

mdh1418 committed:
Address more feedback

lateralusX commented on src/tasks/LibraryBuilder/LibraryBuilder.cs:
```diff
@@ -113,6 +131,17 @@ public override bool Execute()
         GatherAotSourcesObjects(aotSources, aotObjects, extraSources, linkerArgs);
         GatherLinkerArgs(linkerArgs);
 
+        File.WriteAllText(Path.Combine(OutputDirectory, "library_builder.h"),
```
I think we should follow delimitator used by other runtimes header, so library-builder.h

mdh1418 commented on src/mono/msbuild/android/build/AndroidApp.targets:
```diff
@@ -133,6 +140,12 @@
 
+    <PropertyGroup Condition="'$(_IsLibraryMode)' == 'true'">
+      <_EnableUnmanagedCallersOnlyMethodsExport>true</_EnableUnmanagedCallersOnlyMethodsExport>
+      <_UseAotDataFile>$(UseAotDataFile)<_UseAotDataFile>
```
No, I dont have the right logic in autoinit.c to handle `UseAotDataFile == true` right now. Or atleast, it was removed because it wasn't our default use case. I allowed users to specify it if they wanted it and have their own callback logic.

I can add in `mono_install_load_aot_data_hook` thats guarded under an `#if defined(USE_AOT_DATA)` guard. Or should I just remove the flexibility for now?

mdh1418 replied:
Reintroduced the load aot data hook logic all guarded under a definition that only gets included if there are `.aotdata` files in the compiled assemblies, which only occurs if `UseAotDataFile` is enabled when using the LibraryBuilder.

mdh1418 committed:
Add various fixes and make bundle_path a local variable

mdh1418 committed:
Reintroduce load aot data hook in autoinit with guard

mdh1418 committed:
Fix closing tag

mdh1418 committed:
Include guard library-builder.h

mdh1418 commented:
From the Build Analysis, looks like everything is a known test error.
Unfortunately I made a slight typo and I need to update the CMake, but that shouldn't affect any of our tests.

mdh1418 committed:
Fix typo

lateralusX reviewed (approved):
LGTM!

mdh1418 commented:
All failures are known issues.
